### PR TITLE
Add DocumentCollection method to index fields

### PR DESCRIPTION
**Pull request by Gara64**
This PR adds a `indexFields` method to force the indexing of arbitrary fields.

Imagine the following set of documents: 
`[{ planet: "jupiter"}, { planet: "earth", life: true}]`

We want to use the following mango query: `selector: {life: { $exists: false }}`. With the current behaviour, this would try to use an index on `life`, which cannot work as none document with a missing `life` field would enter the index.
This PR make it possible by using `.indexFields(['planet'])` to force an index on another field that the ones queried.

It also: 
* Add a test
* Add docs
* Fix doc for subdirectories (see package.json)

### CODE REVIEW

File: packages/cozy-client/src/queries/dsl.js [hunk 1]
```diff
@@ -19,21 +38,51 @@ export class QueryDefinition {
...
     this.sort = sort
     this.includes = includes
     this.referenced = referenced
     this.limit = limit
     this.skip = skip
   }
 
+  /**
+   * Query documents with a [mango selector](http://docs.couchdb.org/en/latest/api/database/find.html#find-selectors).
+   * Each field passed in the selector will be indexed, except if the indexField is used.
```
Comment (y-lohse) [inline packages/cozy-client/src/queries/dsl.js]:
```suggestion
   * Each field passed in the selector will be indexed, except if the indexField option is used.
```

File: packages/cozy-stack-client/src/DocumentCollection.spec.js [hunk 2]
```diff
@@ -329,6 +329,24 @@ describe('DocumentCollection', () => {
...
+
+    it('should index the specified fields', async () => {
+      const collection = new DocumentCollection('io.cozy.todos', client)
+      await collection.find(
+        { done: { $exists: true } },
+        { indexedFields: ['label'], sort: [{ label: 'desc' }] }
+      )
+      expect(client.fetchJSON).toHaveBeenLastCalledWith(
+        'POST',
+        '/data/io.cozy.todos/_find',
```
Comment (y-lohse) [inline packages/cozy-stack-client/src/DocumentCollection.spec.js]: I don't think this test really tests that the proper fields are indexed. Shouldn't we check for a call to `'/data/io.cozy.todos/_index'`?

File: packages/cozy-client/src/queries/dsl.js [hunk 3]
```diff
@@ -19,21 +38,51 @@ export class QueryDefinition {
...
+   *
+   * @param {Array} fields The fields to return.
+   * @return {QueryDefinition}  The QueryDefinition object.
+   */
   select(fields) {
     return new QueryDefinition({ ...this.toDefinition(), fields })
   }
 
+  /**
+   * Specify which fields should be indexed. This prevent the automatic indexing of the mango fields.
```
Comment (ptbrowne) [inline packages/cozy-client/src/queries/dsl.js]:
```suggestion
   * Specify which fields should be indexed. This prevent the automatic indexing of mango fields.
```

File: packages/cozy-client/src/queries/optimize.js [hunk 4]
```diff
@@ -9,8 +9,8 @@ const isIdQuery = query => query.id || query.ids
 /**
  * Optimize queries on a single doctype
  *
- * @param  {[QueryDefinition]} queries - Queries of a same doctype
- * @return {[QueryDefinition]} Optimized queries
+ * @param  {Array.QueryDefinition} queries - Queries of a same doctype
+ * @return {Array.QueryDefinition} Optimized queries
```
Comment (ptbrowne) [inline packages/cozy-client/src/queries/optimize.js]: I didn't know this syntax (Array.Type) and I cannot find it on the JSX doc. I think it should be `QueryDefinition[]`.

http://usejsdoc.org/tags-param.html#multiple-types-and-repeatable-parameters

**Comment by ptbrowne** [comment]
LGTM

File: packages/cozy-client/src/queries/optimize.js (hunk 4 above)
Comment (Gara64) [reply packages/cozy-client/src/queries/optimize.js]: I got confused by this doc : http://usejsdoc.org/tags-type.html, which is for the type definition. It seems your version is correct

File: packages/cozy-stack-client/src/DocumentCollection.spec.js (hunk 2 above)
Comment (Gara64) [reply packages/cozy-stack-client/src/DocumentCollection.spec.js]: Yes, actually both are necessary. The `_find` call to ensure the sort doesn't include the label, and the `_index` call to make sure the index is created.

**Pull request merged**